### ISSUE TITLE
Race condition on _backtraceLogManager

### ISSUE THREAD

**Issue by vlussenburg** [backtrace-labs/backtrace-unity#28]
In `BacktraceClient.cs`:

```cs
            // set maximum game object depth
            _gameObjectDepth = Configuration.GameObjectDepth == 0
                ? 16 // default maximum game object size
                : Configuration.GameObjectDepth;
            CaptureUnityMessages();
            _reportLimitWatcher = new ReportLimitWatcher(Convert.ToUInt32(Configuration.ReportPerMin));

            _nativeClient = NativeClientFactory.GetNativeClient(Configuration, name);
#if UNITY_2018_4_OR_NEWER

            BacktraceApi = new BacktraceApi(
                credentials: new BacktraceCredentials(Configuration.GetValidServerUrl()),
                ignoreSslValidation: Configuration.IgnoreSslValidation);
#else
            BacktraceApi = new BacktraceApi(new BacktraceCredentials(Configuration.GetValidServerUrl()));
#endif
            _backtraceLogManager = new BacktraceLogManager(Configuration.NumberOfLogs);
```

The effect of `CaptureUnityMessages();` (callback being registered) requires `_backtraceLogManager` to be set.

Fix: make sure `_backtraceLogManager` is initialized before Unity callback is registered.

**Comment by vlussenburg** [comment]
Symptom: these messages in your log:

```
07-03 02:03:06.610 20172 20213 E Unity   : NullReferenceException: Object reference not set to an instance of an object
07-03 02:03:06.610 20172 20213 E Unity   :   at Backtrace.Unity.BacktraceClient.HandleUnityMessage (System.String message, System.String stackTrace, UnityEngine.LogType type) [0x00009] in <d0f5068404a941f186af1a8ab5607519>:0 
07-03 02:03:06.610 20172 20213 E Unity   :   at UnityEngine.Application.CallLogCallback (System.String logString, System.String stackTrace, UnityEngine.LogType type, System.Boolean invokedOnMainThread) [0x00014] in <d71499eab75645c19df09ba17af9a506>:0 
07-03 02:03:06.610 20172 20213 E Unity   : UnityEngine.DebugLogHandler:Internal_Log(LogType, String, Object)
07-03 02:03:06.610 20172 20213 E Unity   : UnityEngine.DebugLogHandler:LogFormat(LogType, Object, String, Object[])
07-03 02:03:06.610 20172 20213 E Unity   : UnityEngine.Logger:Log(LogType, Object)
07-03 02:03:06.610 20172 20213 E Unity   : UnityEngine.Debug:Log(Object)
07-03 02:03:06.610 20172 20213 E Unity   : Backtrace.Unity.BacktraceClient:Refresh()
07-03 02:03:06.610 20172 20213 E Unity   : Backtrace.Unity.BacktraceClient:Awake()
07-03 02:03:06.610 20172 20213 E Unity   :  
07-03 02:03:06.610 20172 20213 E Unity   : (Filename: <d0f5068404a941f186af1a8ab5607519> Line: 0)
07-03 02:03:06.610 20172 20213 E Unity   : 
07-03 02:03:06.610 20172 20213 I Unity   : here4
07-03 02:03:06.610 20172 20213 I Unity   :  
07-03 02:03:06.610 20172 20213 I Unity   : (Filename: ./Runtime/Export/Debug.bindings.h Line: 45)
07-03 02:03:06.610 20172 20213 I Unity   : 
07-03 02:03:06.639 20172 20213 D BacktraceANRWatchdog: Initializing ANR watchdog
07-03 02:03:06.640 20172 20475 D BacktraceANRWatchdog: ANR WATCHDOG - Fri Jul 03 02:03:06 GMT+04:30 2020
07-03 02:03:06.642 20172 20213 E Unity   : NullReferenceException: Object reference not set to an instance of an object
07-03 02:03:06.642 20172 20213 E Unity   :   at Backtrace.Unity.BacktraceClient.HandleUnityMessage (System.String message, System.String stackTrace, UnityEngine.LogType type) [0x00009] in <d0f5068404a941f186af1a8ab5607519>:0 
07-03 02:03:06.642 20172 20213 E Unity   :   at UnityEngine.Application.CallLogCallback (System.String logString, System.String stackTrace, UnityEngine.LogType type, System.Boolean invokedOnMainThread) [0x00014] in <d71499eab75645c19df09ba17af9a506>:0 
07-03 02:03:06.642 20172 20213 E Unity   : UnityEngine.DebugLogHandler:Internal_Log(LogType, String, Object)
07-03 02:03:06.642 20172 20213 E Unity   : UnityEngine.DebugLogHandler:LogFormat(LogType, Object, String, Object[])
07-03 02:03:06.642 20172 20213 E Unity   : UnityEngine.Logger:Log(LogType, Object)
07-03 02:03:06.642 20172 20213 E Unity   : UnityEngine.Debug:Log(Object)
07-03 02:03:06.642 20172 20213 E Unity   : Backtrace.Unity.BacktraceClient:Refresh()
07-03 02:03:06.642 20172 20213 E Unity   : Backtrace.Unity.BacktraceClient:Awake()
07-03 02:03:06.642 20172 20213 E Unity   :  
07-03 02:03:06.642 20172 20213 E Unity   : (Filename: <d0f5068404a941f186af1a8ab5607519> Line: 0)
07-03 02:03:06.642 20172 20213 E Unity   : 
07-03 02:03:06.642 20172 20213 I Unity   : here5
```

**Comment by vlussenburg** [comment]
Tested

- [ ]  Android 10 (emulator)
- [ ]  Android 7 (device)
- [ ] iOS (device)
- [ ] MacOs (Editor)

**Comment by konraddysput** [comment]
@vlussenburg 

this happened because you printed internal logs from the library before we initialized _logManager. Without your changes, this code is working well. In addition to that, we're in a synchronous context so this exception shouldn't happen. 

However, to prevent this situation in the future for other folks, we will move log manager initialization earlier.

Thanks

**Comment by konraddysput** [comment]
Please see https://github.com/backtrace-labs/backtrace-unity/pull/31

**Comment by konraddysput** [comment]
Fix will be available in version 3.0.1 (already on master)

**Comment by vlussenburg** [comment]
Tested master @ 9f298881bb070a7da8ac748a0fe701b45becc4c8

- [x]  Windows Standalone
- [x]  MacOS Editor
- [ ]  MacOs Standalone
- [x]  iOS
- [x]  Android
- [x]  WebGL
